src/clustering/administration/main/command_line.cc

```diff
@@ -645,6 +649,163 @@ service_address_ports_t get_service_address_ports(const std::map<std::string, op
+
+    if (!(key_file && cert_file)) {
+        logERR("must specify both --web-tls-key and --web-tls-cert");
+        return false;
+    }
+
+    return load_tls_key_and_cert(web_tls, *key_file, *cert_file);
+}
+
+bool configure_driver_tls(const std::map<std::string, options::values_t> &opts, SSL_CTX *driver_tls) {
+    boost::optional<std::string> key_file = get_optional_option(opts, "--driver-tls-key");
+    boost::optional<std::string> cert_file = get_optional_option(opts, "--driver-tls-cert");
+    boost::optional<std::string> ca_file = get_optional_option(opts, "--driver-tls-ca");
+
+    if (!(key_file && cert_file)) {
+        logERR("must specify both --driver-tls-key and --driver-tls-cert");
+        return false;
+    }
+
+    if (!load_tls_key_and_cert(driver_tls, *key_file, *cert_file)) {
+        return false;
+    }
+
+    if (ca_file) {
+        if (!SSL_CTX_load_verify_locations(driver_tls, ca_file->c_str(), NULL)) {
+            ERR_print_errors_fp(stderr);
+            return false; 
+        }
+
+        // Mutual authentication.
+        SSL_CTX_set_verify(driver_tls, SSL_VERIFY_PEER|SSL_VERIFY_FAIL_IF_NO_PEER_CERT, NULL);
+    }
+
+    return true;
+}
+
+bool configure_cluster_tls(const std::map<std::string, options::values_t> &opts, SSL_CTX *cluster_tls) {
+    boost::optional<std::string> key_file = get_optional_option(opts, "--cluster-tls-key");
+    boost::optional<std::string> cert_file = get_optional_option(opts, "--cluster-tls-cert");
+    boost::optional<std::string> ca_file = get_optional_option(opts, "--cluster-tls-ca");
+
+    if (!(key_file && cert_file && ca_file)) {
+        logERR("must specify each of --cluster-tls-key, --cluster-tls-cert, and --cluster-tls-ca");
+        return false;
+    }
+
+    if (!load_tls_key_and_cert(cluster_tls, *key_file, *cert_file)) {
+        return false;
+    }
+
+    if (!SSL_CTX_load_verify_locations(cluster_tls, ca_file->c_str(), NULL)) {
+        ERR_print_errors_fp(stderr);
+        return false; 
+    }
+
+    // Mutual authentication.
+    SSL_CTX_set_verify(cluster_tls, SSL_VERIFY_PEER|SSL_VERIFY_FAIL_IF_NO_PEER_CERT, NULL);
+
+    return true;
+}
+
+bool initialize_tls_ctx(SSL_CTX **tls_ctx) {
+    *tls_ctx = SSL_CTX_new(SSLv23_method());
+    if (NULL == *tls_ctx) {
+        ERR_print_errors_fp(stderr);
+        return false;
+    }
+
+    // Only allow TLS v1.2 and prefer server ciphers.
+    SSL_CTX_set_options(*tls_ctx, SSL_OP_NO_SSLv2|SSL_OP_NO_SSLv3|SSL_OP_NO_TLSv1|SSL_OP_NO_TLSv1_1|SSL_OP_CIPHER_SERVER_PREFERENCE);
+
+    /* This is pretty important. We want to use the most secure TLS cipher
+    suite that we can. This simple list only allows ciphers suites which employ
+    ECDHE (Elliptic Curve Diffie-Hellman with Ephemeral keys) for encryption
+    key agreement for Perfect Forward Secrecy and we want to use AESGCM for
+    efficient authenticated encryption. */
+    if (0 == SSL_CTX_set_cipher_list(*tls_ctx, "ECDHE+AESGCM")) {
```

Comment:
`ECDHE-RSA-AES256-GCM-SHA384` is the full cipher that OpenSSL would chose on my machine for TLSv1.2 given the `ECDHE+AESGCM` limitation, lets have a look at all parts.

First off is `ECDHE`, which is the algorithm that is used for the key exchange. The key exchange aims to establish a shared secret, that is a secret known to both the server and the client, but not anyone else, not even if you're listening to the communication between the server and the client. The basic underlying algorithm used is called Diffie-Hellman as @jlhawn mentioned, or `DH` for short. If you happen to be familiar with group theory I strongly recommend reading up on it as it's really neat.

There are two problems with plain Diffie-Hellman though, first of which is that there is an attack given smaller primes and the computing resources reserved to nation states. Two solutions exist, and the first is simply using bigger prime numbers. The current recommendation is to use at least 2048-bit primes which is why the OpenSSL default of 1024-bit I was complaining about before is problematic. The other solution is switching to `ECDH` which is a derived algorithm using Elliptic Curve cryptography instead, which for all intents and purposes I consider magic. Mathematically competent people who have wrapped their heads around this have promised me that it's really neat as well, and safe.

The other problem is that it uses the same very hard to solve problem for every connection, which means that if you solve that problem, hard as it may be, you can decrypt all data over those connections. The trick is to use a different very hard problem for each connection, which is what the last `E`, for Ephemeral as @jlhawn said, is for. This is commonly known as providing forward security.

With that out of the way we're on to `RSA` which is a public-key cryptosystem that's used for authentication in this context. A third party, the Certificate Authority, which both the server and client trust and have the public key off signed a document for the server, and the client, that states that they are who they claim they are. This allows the server and client to prove to each other that they are indeed talking to the system they think they are, which is important.

Next is `AES256` or Advanced Encryption Standard which is a block cipher as @jlhawn said. It takes a piece of data and a secret key and mangles it in such a way that you can't read it unless you have the secret key as well. This is one of the reasons that we needed a secret key to begin with. It works on blocks of 256 bits, and pads the message such that it splits up in to 256 bit chunks. This padding is important, and there are ticks to it I'd rather not get into.

There's a problem with straight AES though, if you give it the same data and key twice, it produces the same mangled output. An attacker could learn that you're sending the same message again, which could be important. To solve this they invented cipher block modes and there's a number of them. The simplest is known as CTR or Counter and simply adds an increasing number to each block, thus now it takes the data, key, and counter to produce the mangled output. `GCM` or Galois/Counter Mode is another such mode and the one that is chosen in this cipher, it's state of the art that combines a counter with an authentication code I will go into in a minute.

To recap, we now know that we're talking to the right computer thanks to authentication via `RSA`, we have a shared secret that only we know thanks to `EDCHE`, are able to send them encrypted messages thanks to `AES256` and we won't send the same encrypted message twice if we have the same data due to `GCM`.

This leaves us with `SHA384`, which is used to provide integrity, allowing the client and server to detect when there's someone in the middle changing the message before they receive them. This is usually done using a MAC or Message Authentication Code, or a HMAC or hash version in particular.

Image you have a cryptographically secure hash algorithm, which is exactly what `SHA384` is. In other words, a function that takes a message and generates a short derived message from it, in such a way that it's nearly impossible to find the original message. You can then take the hash of a message, and a secret key and send that along with the original message. As long as the attacker doesn't have the secret key, at which point you'd be in deep trouble, it can't generate the right hash to send along with the message thus allowing you to check whether it's the original message that you've received. Through clever engineering and some wizardry this is now combined into the block cipher mode `GCM` though, in what's called AEAD or Authenticated Encryption with Associated Data.

Hope this helps, feel free to bug me if anything is unclear.